python_modules/libraries/dagster-dbt/dagster_dbt/dbt_project.py

```diff
@@ -48,16 +49,18 @@ def __init__(
         if not self.project_dir.exists():
             raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
         self._target_folder = Path(target_folder)
-        self._prepare_command = prepare_command
-        # this is ok if it doesn't exist, will get created by `prepare`
+        self._manifest_cmd = generate_artifacts_command
+        # this is ok if it doesn't exist, will get created during sync
         self._packaged_project_dir = Path(packaged_project_dir) if packaged_project_dir else None
 
         if self._should_use_built_artifacts():
             self._manifest_path = self._current_project_dir.joinpath(
                 self._target_folder, "manifest.json"
             )
         else:
-            self._manifest_path = self._prepare_manifest()
+            self._manifest_path = self.generate_artifacts(
+                log_level=logging.DEBUG,  # quiet logging for init time generation
+            )
```

Comment:
[2]

Reply:
init time generation for local development can be handled with the `DAGSTER_CLI_DEV` flag right?

Reply:
its already gated by that env var here, maybe I am not following what you mean. 

What i meant is that a scheme where `dbt` / `dbt_fn` are arguments to `generate_artifacts` it will either be invoked differently in the local dev flow or that generation call has to get lifted back out of here in to user code

Reply:
Ah, I didn't realize that the artifacts were generated at init time of the `DbtProject` object.

Can we just defer the artifact creation when the `DbtProject` is actually taken up by `@dbt_assets`?

Reply:
we can defer, thats how i had it before PR feedback to do it init time https://github.com/dagster-io/dagster/pull/19923#discussion_r1499767914

but even if its deferred, you would still need to bind the custom resource or function before it gets called somehow, assuming a setup outlined here https://github.com/dagster-io/dagster/pull/20632#pullrequestreview-1955194246